### PR TITLE
feat($httpBackend): flush requests in desired order

### PR DESCRIPTION
**Pull request by dizel3d**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

feature

**What is the current behavior? (You can also link to an open issue here)**
#13717 We are currently unable to test what happens if two HTTP requests complete in a different order with the current API.

**What is the new behavior (if this is a feature change)?**

Now `$httpBackend.flush` has the second parameter `start`. So you can flush any range of pending requests array. Hence you can flush requests in a different order.

**Does this PR introduce a breaking change?**

No.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

It was impossible to flush pending requests in desired order.
Now it is possible because you can specify a `start` number.

### CODE REVIEW

File: src/ngMock/angular-mocks.js [hunk 1]
```diff
@@ -1788,24 +1788,29 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
    * @ngdoc method
    * @name $httpBackend#flush
    * @description
-   * Flushes all pending requests using the trained responses.
+   * Flushes pending requests using the trained responses in the order they arrived.
```
Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: I would mention the `start/skip` parameter in the description.

File: src/ngMock/angular-mocks.js [hunk 2]
```diff
@@ -1788,24 +1788,29 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
...
+   * If there are no pending requests when the flush method is called
+   * an exception is thrown (as this typically a sign of programming error).
    *
-   * @param {number=} count Number of responses to flush (in the order they arrived). If undefined,
-   *   all pending requests will be flushed. If there are no pending requests when the flush method
-   *   is called an exception is thrown (as this typically a sign of programming error).
+   * @param {number=} count Number of responses to flush. If undefined,
+   *   all pending requests will be flushed beginning at `start`.
+   * @param {number=} [start=0] Sequential request number at which to begin to flush.
```
Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: I think `skip` is a better name for this param. It is more intuitive imo.

(Also, have you checked how the resulting docs look like? Can dgeni really understand the `[<name>=<default value>]` syntax?)

File: src/ngMock/angular-mocks.js [hunk 3]
```diff
@@ -1788,24 +1788,29 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
...
    */
-  $httpBackend.flush = function(count, digest) {
+  $httpBackend.flush = function(count, start, digest) {
     if (digest !== false) $rootScope.$digest();
     if (!responses.length) throw new Error('No pending request to flush !');
 
+    start = angular.isDefined(start) && start !== null ? start : 0;
```
Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: Nitpick: `start = start || 0` should work just as well.

File: src/ngMock/angular-mocks.js [hunk 4]
```diff
@@ -1788,24 +1788,29 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
...
-   *   all pending requests will be flushed. If there are no pending requests when the flush method
-   *   is called an exception is thrown (as this typically a sign of programming error).
+   * @param {number=} count Number of responses to flush. If undefined,
+   *   all pending requests will be flushed beginning at `start`.
+   * @param {number=} [start=0] Sequential request number at which to begin to flush.
    */
-  $httpBackend.flush = function(count, digest) {
+  $httpBackend.flush = function(count, start, digest) {
     if (digest !== false) $rootScope.$digest();
     if (!responses.length) throw new Error('No pending request to flush !');
```
Comment (gkalpak) [inline src/ngMock/angular-mocks.js]: You need to account for `start` here as well.

File: test/ngMock/angular-mocksSpec.js [hunk 5]
```diff
@@ -1532,6 +1532,38 @@ describe('ngMock', function() {
...
+
+        hb.when('GET').respond(200, '');
+        hb('GET', '/some', null, dontCallMe);
+        hb('GET', '/some', null, callback);
+        hb('GET', '/some', null, callback);
+        hb('GET', '/some', null, dontCallMe);
+
+        hb.flush(2, 1);
+        expect(dontCallMe).not.toHaveBeenCalled();
+        expect(callback).toHaveBeenCalled();
```
Comment (gkalpak) [inline test/ngMock/angular-mocksSpec.js]: This is redundant.

File: test/ngMock/angular-mocksSpec.js [hunk 6]
```diff
@@ -1532,6 +1532,38 @@ describe('ngMock', function() {
...
+
+        hb.when('GET').respond(200, '');
+        hb('GET', '/some', null, dontCallMe);
+        hb('GET', '/some', null, dontCallMe);
+        hb('GET', '/some', null, callback);
+        hb('GET', '/some', null, callback);
+
+        hb.flush(null, 2);
+        expect(dontCallMe).not.toHaveBeenCalled();
+        expect(callback).toHaveBeenCalled();
```
Comment (gkalpak) [inline test/ngMock/angular-mocksSpec.js]: This is redundant.

**Comment by gkalpak** [comment]
I think this is a nice addition. Left some (mostly) minor comments.
@dizel3d, feel free to ping me when/if you have addressed them, so I can take another look.

File: src/ngMock/angular-mocks.js (hunk 2 above)
Comment (dizel3d) [inline src/ngMock/angular-mocks.js]: Yes, default value looks good in doc. Also this syntax is already in use in other parts of the project.

File: test/ngMock/angular-mocksSpec.js (hunk 5 above)
Comment (dizel3d) [inline test/ngMock/angular-mocksSpec.js]: Yes it is.
(I just noticed `toHaveBeenCalled` together with `toHaveBeenCalledTimes` in test above.)

File: test/ngMock/angular-mocksSpec.js (hunk 6 above)
Comment (dizel3d) [inline test/ngMock/angular-mocksSpec.js]: Yes, you are right.

File: src/ngMock/angular-mocks.js (hunk 1 above)
Comment (dizel3d) [inline src/ngMock/angular-mocks.js]: Ok, I'll do.
(I am not a native speaker so I will be thankful to get any idea about the text.)

**Comment by dizel3d** [comment]
@gkalpak, I've made changes.

**Comment by Narretz** [comment]
This introduces a breaking change as "skip" is now the second argument, no?

**Comment by dizel3d** [comment]
@Narretz, do you mean due to the changes `digest` became 3rd argument? Yes, it did. But `digest` is undocumented and used internally only.

**Comment by gkalpak** [comment]
Only `count` was a documented parameter. The previously second (now third) `digest` parameter was undocumented/private, so nobody should be using it (except us). I don't consider this a breaking change technically, but it can break apps that are trying to be smart and take advantage of undocumented/private stuff.

Since this is not critical (i.e. it is an extra feature), I am fine landing it in 1.6 only (but I wouldn't mark it as a breaking change).

**Comment by gkalpak** [comment]
LGTM (as long as Travis is happy).
@Narretz, what do you think about BC/landing branch?

**Comment by Narretz** [comment]
Okay, since the digest parameter is private I'm fine with landing this in 1.5.
Btw, this is a duplicate of https://github.com/angular/angular.js/pull/6412

Can you add
`Closes #13717` 
at the end of the commit message?

**Comment by dizel3d** [comment]
@Narretz, actually the PR closes only paragraph 2 of #13717.
(The PR contains 2 commits. In which should I add `Closes #13717`?)

**Comment by Narretz** [comment]
@dizel3d Then just add `Related #13717` to any of the 2 commits. We'll squash them together on merge.

**Comment by dizel3d** [comment]
@Narretz Done!

**Comment by Narretz** [comment]
LGTM

**Pull request closed without merging**